Add pull requests guide doc (#2145)
dotnet/runtime: docs/pr-guide.md

If during the code review merge conflict occurs the area owner is responsible for its resolution. Pull request should not be on hold due to author unwillingness to resolve code conflicts. GitHub makes this easier by allowing simple conflict resolution using [conflict-editor](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github).

## Merging Pull Requests
Anyone with write access can merge a pull request manually or by setting the [auto-merge](/labels/auto-merge) label when it satisfies all of the following conditions:

* The PR has been approved by at least one reviewer and any other objections are addressed.
 * You can request another review from the original reviewer.
* The PR successfully builds and passes all tests in the Continuous Integration (CI) system.

>  * You can trigger a rebuild by commenting `/arp run` or manually re-run only the failing lanes in Azure DevOps tab.

Reviewer: This should be `/azp run <pipeline-name>`. 

See my comment:
> Actually it is /azp run <pipeline name> (i.e /azp run runtime). In fact, we don't recommend /azp run as it triggers ALL the outerloop and non-required builds and we're working with AzDo on providing a feature were we can disable that behavior.

Author: I don't think that's helpful unless we explain somewhere what <pipeline-name> is.

Reviewer: Why it wouldn't be helpful saying people shouldn't use `/azp run`? And what I mean, is that `arp run` is not correct it is `/azp run`. 

```suggestion
 * You can trigger a rebuild by commenting `/azp run <pipeline name>` or manually re-run only the failing lanes in Azure DevOps tab.
```

Reviewer 2: @safern I think he's asking how to get a list of pipelines. Eg., /azp help or something,

Reviewer: I see, yeah, then we can mention people can get the list of pipelines with `/azp list`

>  * Reach out to the infrastructure team for assistance on Teams [channel](https://teams.microsoft.com/l/channel/19%3ab27b36ecd10a46398da76b02f0411de7%40thread.skype/Infrastructure?groupId=014ca51d-be57-47fa-9628-a15efcc3c376&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) in other cases.

Reviewer 2: Might have got lost in another comment. This is only accessible to folks authenticated into Microsoft corpnet. I don't think it matters if we put the link here, but it's probably worth a note explaining that community members can't access it. In the future, it might be interesting to have a channel that we all can access.

Reviewer 3: Maybe say community members can ask on Gitter, since that's open to everyone?

Reviewer 2: Yeah, that may be better

# Pull Request Guide

## Contributing Rules
All contributions to dotnet/runtime repo are made via pull requests (PRs) rather than through direct commits. The pull requests are reviewed and merged by the repository maintainers after a review and approval from at least one area maintainer.

> To merge pull requests, you must have write permissions in the repository. If you are a member of the .NET team or have an official partnership you can request the access [here](https://repos.opensource.microsoft.com/dotnet/teams/dotnet-corefx/join/).

Reviewer 3: ```suggestion
To merge pull requests, you must have write permissions in the repository. If you are a member of the .NET team or have an official partnership you can request access [here](https://repos.opensource.microsoft.com/dotnet/teams/dotnet-corefx/join/).
```

## Quick Code Review Rules

> * Do not mix unrelated changes in one pull request. For example, a code style changes should never be mixed with a bug fix.

Reviewer 3: ```suggestion
* Do not mix unrelated changes in one pull request. For example, a code style change should never be mixed with a bug fix.
```

> * All changes should follow existing code style. You can read more about different code styles at [docs/coding-guidelines](coding-guidelines/).

Reviewer 3: ```suggestion
* All changes should follow the existing code style. You can read more about different code styles at [docs/coding-guidelines](coding-guidelines/).
```

> * Use Draft pull request for changes you are still working on but want early CI loop feedback. When you think your changes are ready for review [change status](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request) of your pull request.

Reviewer 3: ```suggestion
* Use Draft pull requests for changes you are still working on but want early CI loop feedback. When you think your changes are ready for review, [change the status](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request) of your pull request.
```

* Avoid rebasing your changes. If you are asked to make changes during the review process do them as a new commit.

## Pull Request Ownership

Every pull request will have automatically a single `area-*` label assigned. The label not only indicates the code segment which the change touches but also the owner. We maintain a list of [areas owners](area-owners.md) for all dotnet/runtime labels. They are responsible for landing pull requests in their area in a timely manner and for helping contributors with their submitted pull request. You can ask them for assistance if you need help with landing your changes.

> If during the code review merge conflict occurs the area owner is responsible for its resolution. Pull request should not be on hold due to author unwillingness to resolve code conflicts. GitHub makes this easier by allowing simple conflict resolution using [conflict-editor](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github).

Reviewer 3: ```suggestion
If during the code review process a merge conflict occurs the area owner is responsible for its resolution. Pull requests should not be on hold due to the author's unwillingness to resolve code conflicts. GitHub makes this easier by allowing simple conflict resolution using the [conflict-editor](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github).
```

## Merging Pull Requests
Anyone with write access can merge a pull request manually or by setting the [auto-merge](/labels/auto-merge) label when it satisfies all of the following conditions:

* The PR has been approved by at least one reviewer and any other objections are addressed.

>  * You can request another review from the original reviewer

Reviewer 3: ```suggestion
 * You can request another review from the original reviewer.
```

* The PR successfully builds and passes all tests in the Continuous Integration (CI) system.

>  * You can trigger rebuild using `/arp run` comment or manually re-run only failing lanes in Azure DevOps tab.

Reviewer 3: ```suggestion
 * You can trigger a rebuild by commenting `/arp run` or manually re-run only the failing lanes in Azure DevOps tab.
```

Reviewer 2: @safern is `/arp run` what we recommend?

Reviewer: Actually it is `/azp run <pipeline name>` (i.e `/azp run runtime`). In fact, we don't recommend `/azp run` as it triggers ALL the outerloop and non-required builds and we're working with AzDo on providing a feature were we can disable that behavior.

>  * Reach out to infrastructure team for assistance on Teams [channel](https://teams.microsoft.com/l/channel/19%3ab27b36ecd10a46398da76b02f0411de7%40thread.skype/Infrastructure?groupId=014ca51d-be57-47fa-9628-a15efcc3c376&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) in other cases.

Reviewer 3: ```suggestion
 * Reach out to the infrastructure team for assistance on Teams [channel](https://teams.microsoft.com/l/channel/19%3ab27b36ecd10a46398da76b02f0411de7%40thread.skype/Infrastructure?groupId=014ca51d-be57-47fa-9628-a15efcc3c376&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) in other cases.
```

Reviewer 2: This Teams channel is only accessible from Microsoft corpnet, I think?

 * In rare cases the pull request can be merged with failing tests if the person merging it also creates tracking issue for flaky tests.

> Please always squash the pull request unless explicitly requested by the PR author. Do so, even if the PR contains only one commit. It creates a simpler history than a Merge Commit.

Reviewer 3: ```suggestion
Please always squash the pull request unless explicitly requested by the PR author, even if the PR contains only one commit. It creates a simpler history than a Merge Commit.
```

Reviewer 3: [This](https://github.com/dotnet/runtime/blob/master/docs/issues-pr-management.md#is-this-a-squash-on-merge-repo) should probably also be updated alongside this comment, since it's currently unclear (presumably the snippet was copy-pasted and made sense in context wherever it came from).

> Please always squash the pull request unless explicitly requested by the PR author. Do so, even if the PR contains only one commit. It creates a simpler history than a Merge Commit.

Reviewer 2: ```suggestion
Please always squash the pull request unless there are special circumstances. Do so, even if the PR contains only one commit. It creates a simpler history than a Merge Commit. "Special circumstances" are rare, and typically mean that there are a series of cleanly separated changes that will be too hard to understand if squashed together, or for some reason we want to preserve the ability to bisect them.
```

Ie., - request of the PR author is _not_ sufficient reason to not squash. (Some contributors prefer their changes not be squashed, but this is not repo policy.)

## Blocking Pull Request Merging

> If for whatever reason you would like to move your pull request back to in-progress status to avoid merging it in the current form. You can do that by adding [WIP] prefix to the pull request title.

Reviewer 3: ```suggestion
If for whatever reason you would like to move your pull request back to an in-progress status to avoid merging it in the current form, you can do that by adding [WIP] prefix to the pull request title.
```
